[REEF-498] Create scripts for release versioning, packaging, and checksum

dongjun-Lee opened the pull request:
This pull request addressed the issue by
* creating `release.py` which creates a release candidate and relevant files.
* creating `change_version.py` which changes versions in every pom.xml and relevant files.

JIRA: [REEF-498](https://issues.apache.org/jira/browse/REEF-498)

markusweimer commented on dev/change_version.py:
```diff
@@ -0,0 +1,142 @@
+            break
+        changed_str += line
+
+    r = re.compile('<version>(.*?)</version>')
```
Is this safe? Worst case, this picks up additional lines in the pom. Can we parse the XML instead?

dongjun-Lee replied:
@markusweimer Thank you for your comment. At first, I parsed XML and rebuilt it but it ignored comments and made unnecessary changes such as indent or newline. I think this is quite safe because change happens only once for a file.

markusweimer commented on dev/change_version.py:
```diff
@@ -0,0 +1,142 @@
+    f.close()
+
+"""
+Change the name of shaded.jar in lang/cs/Org.Apache.REEF.Client/run.cmd
```
I believe `run.cmd` is deprecated. @jwang98052, correct?

jwang98052 replied:
It is still used when running REEF.Tests on Yarn.

markusweimer commented on dev/change_version.py:
```diff
@@ -0,0 +1,142 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
This script also needs to change the versions found in `AssemblyInfo.cs` as well as the NuGet versions.

dongjun-Lee commented:
I changed the script to change build.props and every AssemblyInfo.cs files. Since parsing XMLs and rebuilding have problems like ones I've mentioned above, instead of changing the parsing process I've changed the existing method to be safer.
@markusweimer @dafrista , could you please take a look at this commit? Thank you.

dafrista commented on dev/release.py:
```diff
@@ -0,0 +1,161 @@
+            md5_file = open(file_name + ".md5", "w")
+            md5_file.write("MD5(" + file_name + ")= " + md5 + "\n")
+            sha_file = open(file_name + ".sha", "w")
+            sha_file.write("SHA(" + file_name + ")= " + sha + "\n")
```
It looks like the output is meant to follow `openssl`. But it's not quite the same. Replacing `SHA(` with `SHA512(` will do that.

(We're still discussing at dev@ whether following `openssl` should be the correct output format.)

dafrista replied:
If there are no objections, let's change to "coreutils style". This is the style [currently used on rc2](https://dist.apache.org/repos/dist/dev/incubator/reef/0.12.0-incubating-rc2/), the checksum followed by a space, asterisk, then the filename. (The asterisk means that the file being checksum-ed was treated as binary, which is what we want.)

markusweimer replied:
+1 on coreutils. They are widely available and we can make PowerShell equivalents

dongjun-Lee commented:
@dafrista Thank you for the comment. I addressed it.

markusweimer commented on dev/release.py:
```diff
@@ -0,0 +1,161 @@
+    sha = open(file_name + ".sha512.txt").read().split(" ")[0]
+    return_str += "SHA: " + sha + "\n"
+
+    return_str += "\nRelease artifacts are signed with the key. The KEYS file is available here:\n"
```
`the key` sounds odd here. Maybe a better formulation would be:

```
Release artifacts are signed with a key found in the KEYS file available here:
```

dongjun-Lee commented:
Thanks @markusweimer, I fixed it.

dafrista commented on dev/change_version.py:
```diff
@@ -0,0 +1,265 @@
+If you use the option "-s false", bulid.props changes as,
+ <RemoveIncubating>true</RemoveIncubating>
+ <IsSnapshot>false</IsSnapshot>
+ <SnapshotNumber>0</SnapshotNumber>
```
@markusweimer the snapshot uses the `##` format (two digits), right? If so, we should change this to show e.g. 00.

markusweimer replied:
I believe so. We need to generate a unique version each time we want to publish a NuGet SNAPSHOT. 99 snapshots between releases seems more realistic than 9 :-)

dafrista commented on dev/change_version.py:
```diff
@@ -0,0 +1,265 @@
+            changed_str += line.replace(old_is_snapshot, is_snapshot)
+        elif "<SnapshotNumber>" in line and "</SnapshotNumber>" in line and is_snapshot=="false":
+            old_snapshot_number = r3.search(line).group(1)
+            changed_str += line.replace(old_snapshot_number, "0")
```
See above comment. This might have to be `"00"`.

jwang98052 replied:
Yes. Unique and allow up to 99 was the original reason.

dongjun-Lee commented:
@dafrista I fixed it. Thank you for the comment.

markusweimer commented:
Given that we've done the release, is it time to merge this?

dafrista commented:
Thanks for asking!

This has been a classic example of changing specifications... and I'd like to ask for one more. When changing version numbers for the maven release, *only* the poms must be changed. I accomplished this by commenting out stuff in the script. But, it would make sense to add such an option. @dongjun-Lee is this possible? Thanks for bearing with me.

Once that gets in, we can merge. Then I'll change the cwiki entries to add how some of the steps can be made easy via the scripts.

dongjun-Lee commented:
Thank you @dafrista . I added an option to change only the poms.

dafrista commented:
@dongjun-Lee I've updated the [release manager guide](https://cwiki.apache.org/confluence/display/REEF/Release+Manager+Guide). Please double-check the script instructions that I've added.

dongjun-Lee commented:
@dafrista Sorry, I made a mistake. I fixed the bug.
Thanks a lot for the guide. In the guide, Step 3 might be included in step 4 because `release.py` includes removing "-SNAPSHOT" and changing versions in all files.

dafrista commented:
@dongjun-Lee thanks for the update! It looks good.

Can we remove the use of `change_version` in `release`?

I'm probably contradicting an earlier request. However, I now think this amount of automation is dangerous. The release manager should tarball the exact changes he made on the branch.

After this final change, I'm happy to merge.

dafrista commented on dev/release.py:
```diff
@@ -0,0 +1,161 @@
+import fnmatch
+import tarfile
+import hashlib
+import change_version
```
Let's remove this (more details in the previous comment).

dongjun-Lee commented:
@dafrista I removed it. Thank you :)
